### PR TITLE
Adds additional kwargs to Documenter plugin

### PR DESCRIPTION
**Pull request by morris25**

### CODE REVIEW

File: src/plugins/documenter.jl [hunk 1]
```diff
@@ -32,6 +32,19 @@ function gen_plugin(p::Documenter, t::Template, pkg_name::AbstractString)
...
 
+    kwargs_string = if :additional_kwargs in fieldnames(typeof(p))
+        set_kwargs = ["modules", "format", "pages", "repo", "sitename", "authors", "assets"]
+
+        # We want something that looks like the following:
+        #     key1="val1",
+        #     key2="val2",
+        #
+        kwargs = (x for x in p.additional_kwargs if first(x) ∉ set_kwargs)
+        join(string(tab, first(p), "=", repr(last(p)), ",\n") for p in kwargs)
```
Comment (omus) [inline src/plugins/documenter.jl]: Split up pair earlier to make code easier to read:
```julia
(k for (k, v) in p.additional_kwargs if k ∉ set_kwargs)
```

File: src/plugins/documenter.jl [hunk 2]
```diff
@@ -32,6 +32,19 @@ function gen_plugin(p::Documenter, t::Template, pkg_name::AbstractString)
         "[]"
     end
 
+    kwargs_string = if :additional_kwargs in fieldnames(typeof(p))
+        set_kwargs = ["modules", "format", "pages", "repo", "sitename", "authors", "assets"]
```
Comment (omus) [inline src/plugins/documenter.jl]: Maybe rename to `standard_kwargs` or `ignored_kwargs`

File: src/plugins/documenter.jl (hunk 1 above)
Comment (omus) [reply src/plugins/documenter.jl]: There probably should also be a warning if a user accidentally included one of these keywords in the `additional_kwargs` field

File: test/tests.jl [hunk 3]
```diff
@@ -13,6 +13,11 @@ end
 struct Bar <: CustomPlugin end
 # A dummy Plugin subtype.
 struct Baz <: Plugin end
+# A Documenter with extra kwargs
+struct Bat <: Documenter
```
Comment (omus) [inline test/tests.jl]: `Qux` is the next in line: https://en.wikipedia.org/wiki/Metasyntactic_variable

**Comment by omus** [comment]
@christopher-dG any comments?

**Comment by christopher-dG** [comment]
I'll review shortly

File: src/plugins/documenter.jl [hunk 4]
```diff
@@ -32,6 +32,27 @@ function gen_plugin(p::Documenter, t::Template, pkg_name::AbstractString)
...
     end
 
+    kwargs_string = if :additional_kwargs in fieldnames(typeof(p))
+        standard_kwargs = ["modules", "format", "pages", "repo", "sitename", "authors", "assets"]
+
+        # We want something that looks like the following:
+        #     key1="val1",
+        #     key2="val2",
+        #
+        valid_keys = [k for k in keys(p.additional_kwargs) if k ∉ standard_kwargs]
```
Comment (christopher-dG) [inline src/plugins/documenter.jl]: Any chance we could use `filter(k -> !in(k, standard_kwargs), keys(p.additional_kwargs))`?
I prefer to use map/filter over comprehensions and I'm not very fond of using unicode symbols.

File: src/plugins/documenter.jl [hunk 5]
```diff
@@ -32,6 +32,27 @@ function gen_plugin(p::Documenter, t::Template, pkg_name::AbstractString)
...
+        if length(p.additional_kwargs) > length(valid_keys)
+            invalid_keys = (repr(k) for k in keys(p.additional_kwargs) if k in standard_kwargs)
+            @warn string(
+                "Ignoring predefined Documenter kwargs ",
+                join(invalid_keys, ", "),
+                " from additional kwargs."
```
Comment (christopher-dG) [inline src/plugins/documenter.jl]: None of the log messages have punctuation, so maybe let's remove the period at the end here

File: test/tests.jl [hunk 6]
```diff
@@ -397,4 +397,32 @@ end
...
+    pkg_dir = joinpath(t.dir, test_pkg)
+    p = Qux(
+        [],
+        Dict("strict"=>true, "checkdocs"=>:none, "format"=>:markdown, "stringarg"=>"string")
+    )
+    warn_str = "Ignoring predefined Documenter kwargs \"format\" from additional kwargs."
+    @test_logs (:warn, warn_str) gen_plugin(p, t, test_pkg)
+    make = readchomp(joinpath(pkg_dir, "docs", "make.jl"))
+    @test occursin(
+        strip("""
```
Comment (christopher-dG) [inline test/tests.jl]: Could we `lstrip` this string to make sure that the keyword list has a trailing newline?

File: src/plugins/documenter.jl [hunk 7]
```diff
@@ -32,6 +32,27 @@ function gen_plugin(p::Documenter, t::Template, pkg_name::AbstractString)
...
+            )
+        end
+        join(string(tab, k, "=", repr(p.additional_kwargs[k]), ",\n") for k in valid_keys)
```
Comment (christopher-dG) [inline src/plugins/documenter.jl]: I'd prefer `map(k -> string(tab, k, "=", repr(p.additional_kwargs[k]), ",\n"), valid_keys)` here too

File: src/plugins/documenter.jl [hunk 8]
```diff
@@ -32,6 +32,27 @@ function gen_plugin(p::Documenter, t::Template, pkg_name::AbstractString)
         "[]"
     end
 
+    kwargs_string = if :additional_kwargs in fieldnames(typeof(p))
+        standard_kwargs = ["modules", "format", "pages", "repo", "sitename", "authors", "assets"]
```
Comment (christopher-dG) [inline src/plugins/documenter.jl]: Perhaps make this a `const` at the top of the file? Also if we name it something like `standard_kws` if fits in 92 characters 🙃

**Comment by christopher-dG** [comment]
One more thing we could do is check the `fieldtype` of `additional_kwargs` to make sure that it's not going to break.

File: test/tests.jl [hunk 9]
```diff
@@ -397,4 +397,52 @@ end
...
+end
+
+@testset "Documenter add kwargs" begin
+    t = Template(; user=me)
+    pkg_dir = joinpath(t.dir, test_pkg)
+
+    function check_kwargs(kwargs, warn_str)
+        p = Qux(
+            [],
+            kwargs
```
Comment (christopher-dG) [inline test/tests.jl]: Trailing comma!

File: src/plugins/documenter.jl [hunk 10]
```diff
@@ -32,6 +46,27 @@ function gen_plugin(p::Documenter, t::Template, pkg_name::AbstractString)
         "[]"
     end
 
+    kwargs_string = if in(:additional_kwargs, fieldnames(typeof(p))) &&
```
Comment (christopher-dG) [inline src/plugins/documenter.jl]: `x in y` is a bit prettier

File: test/tests.jl (hunk 9 above)
Comment (christopher-dG) [reply test/tests.jl]: actual upon further review this can definitely be on one line.

**Pull request merged**